torchdynamo/_eval_frame.c

```diff
@@ -390,52 +386,49 @@ static PyObject *_custom_eval_frame_run_only(PyThreadState *tstate,
   }
 }
 
+static int active_dynamo_threads = 0;
+
+static PyObject *increment_working_threads(PyThreadState *tstate) {
+  active_dynamo_threads = active_dynamo_threads + 1;
+  if (active_dynamo_threads > 0) {
+    enable_eval_frame_shim(tstate);
+  }
+  Py_RETURN_NONE;
+}
+
+static PyObject *decrement_working_threads(PyThreadState *tstate) {
+  if (active_dynamo_threads > 0) {
+    active_dynamo_threads = active_dynamo_threads - 1;
+    if (active_dynamo_threads == 0) {
+      enable_eval_frame_default(tstate);
+    }
+  }
+  Py_RETURN_NONE;
+}
+
 static PyObject *set_eval_frame(PyObject *new_callback, PyThreadState *tstate) {
   // Change the eval frame callback and return the old one
   //  - None: disables TorchDynamo
   //  - False: run-only mode (reuse existing compiles)
-  //  - Python callable(): enables TorchDynamo
-
-  PyObject *old_callback;
-#if PY_VERSION_HEX >= 0x03090000
-  void *old_eval_frame = _PyInterpreterState_GetEvalFrameFunc(tstate->interp);
-#else
-  void *old_eval_frame = tstate->interp->eval_frame;
-#endif
-  if (old_eval_frame == &custom_eval_frame) {
-    old_callback = eval_frame_callback_get();
-  } else if (old_eval_frame == &custom_eval_frame_run_only) {
-    old_callback = Py_False;
-  } else if (old_eval_frame == &_PyEval_EvalFrameDefault) {
-    old_callback = Py_None;
-  } else {
-    CHECK(false);
-  }
-
+  //  - Python callable(): enables TorchDynamo    
+  PyObject *old_callback = eval_frame_callback_get();
+  
   // owned by caller
   Py_INCREF(old_callback);
-
+    
   if (new_callback == Py_None) {
-    disable_eval_frame(tstate);
-  } else if (new_callback == Py_False) {
-    enable_run_only_eval_frame(tstate);
+    decrement_working_threads(tstate);
   } else {
-    enable_eval_frame(tstate);
-#ifdef TORCHDYNAMO_MULTI_THREAD
-  } // callback is private, so always clear it
-#endif
+    increment_working_threads(tstate);
```

Comment:
I'm no expert in cpython reference counting, but I think this is correct. We increment the new_callback reference since we take it and store it, and we lower the old one. The shims don't need ref counting, and nothing else really changed in terms of crossing the c/py boundary. LMK if I did something clowny and overlooked it though!

Reply:
Thanks yeah.  I didn't see any issue, but it is subtle enough I want to double check.